// nativescript-ng-shadow/angular/ng-shadow.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var platform_1 = require("tns-core-modules/platform");
var shadow_1 = require("../common/shadow");
var NativeShadowDirective = /** @class */ (function () {
    function NativeShadowDirective(el) {
        this.el = el;
        this.loaded = false;
        this.initialized = false;
    }
    NativeShadowDirective.prototype.ngOnInit = function () {
        this.initializeCommonData();
        if (platform_1.isAndroid) {
            this.initializeAndroidData();
        }
        else if (platform_1.isIOS) {
            this.initializeIOSData();
        }
        if (this.shadow && this.shadow.elevation) {
            if (platform_1.isAndroid) {
                this.loadFromAndroidData(this.shadow);
            }
            else if (platform_1.isIOS) {
                this.loadFromIOSData(this.shadow);
            }
        }
        this.applyShadow();
        this.initialized = true;
    };
    NativeShadowDirective.prototype.onLoaded = function () {
        this.loaded = true;
        // Weirdly ngOnInit isn't called on iOS on demo app
        // Managed to get it working on iOS when applying to
        // FlexboxLayout, but on the demo app, we apply to a
        // Label, and, for that case, ngOnInit isn't called
        // This is just enforcing the Directive is initialized
        // before calling this.applyShadow()
        if (!this.initialized) {
            this.ngOnInit();
        }
        this.applyShadow();
    };
    NativeShadowDirective.prototype.ngOnChanges = function (changes) {
        if (this.loaded &&
            !!changes &&
            (changes.hasOwnProperty('shadow') ||
                changes.hasOwnProperty('elevation') ||
                changes.hasOwnProperty('shape') ||
                changes.hasOwnProperty('bgcolor') ||
                changes.hasOwnProperty('cornerRadius') ||
                changes.hasOwnProperty('translationZ') ||
                changes.hasOwnProperty('maskToBounds') ||
                changes.hasOwnProperty('shadowColor') ||
                changes.hasOwnProperty('shadowOffset') ||
                changes.hasOwnProperty('shadowOpacity') ||
                changes.hasOwnProperty('shadowRadius'))) {
            if (changes.hasOwnProperty('shadow') &&
                !changes.hasOwnProperty('elevation') &&
                typeof changes.shadow.currentValue === "number") {
                this.elevation = changes.shadow.currentValue;
            }
            if (changes.shadow && changes.shadow.currentValue.elevation) {
                if (platform_1.isAndroid) {
                    this.loadFromAndroidData(this.shadow);
                }
                else if (platform_1.isIOS) {
                    this.loadFromIOSData(this.shadow);
                }
            }
            this.applyShadow();
        }
    };
    NativeShadowDirective.prototype.applyShadow = function () {
        if (this.shadow === null ||
            this.shadow === undefined ||
            (this.shadow === '' && !this.elevation)) {
            return;
        }
        shadow_1.Shadow.apply(this.el.nativeElement, {
            elevation: this.elevation,
            shape: this.shape,
            bgcolor: this.bgcolor,
            cornerRadius: this.cornerRadius,
            translationZ: this.translationZ,
            maskToBounds: this.maskToBounds,
            shadowColor: this.shadowColor,
            shadowOffset: this.shadowOffset,
            shadowOpacity: this.shadowOpacity,
            shadowRadius: this.shadowRadius,
        });
    };
    NativeShadowDirective.prototype.initializeCommonData = function () {
        var tShadow = typeof this.shadow;
        if ((tShadow === 'string' || tShadow === 'number') && !this.elevation) {
            this.elevation = this.shadow ?
                parseInt(this.shadow, 10) : 2;
        }
        var tElevation = typeof this.elevation;
        if (tElevation === 'string' || tElevation === 'number') {
            this.elevation = this.elevation ?
                parseInt(this.elevation, 10) : 2;
        }
    };
    NativeShadowDirective.prototype.initializeAndroidData = function () {
        if (typeof this.cornerRadius === 'string') {
            this.cornerRadius = parseInt(this.cornerRadius, 10);
        }
        if (typeof this.translationZ === 'string') {
            this.translationZ = parseInt(this.translationZ, 10);
        }
    };
    NativeShadowDirective.prototype.initializeIOSData = function () {
        if (typeof this.shadowOffset === 'string') {
            this.shadowOffset = parseFloat(this.shadowOffset);
        }
        if (typeof this.shadowOpacity === 'string') {
            this.shadowOpacity = parseFloat(this.shadowOpacity);
        }
        if (typeof this.shadowRadius === 'string') {
            this.shadowRadius = parseFloat(this.shadowRadius);
        }
    };
    NativeShadowDirective.prototype.loadFromAndroidData = function (data) {
        this.elevation = data.elevation || this.elevation;
        this.shape = data.shape || this.shape;
        this.bgcolor = data.bgcolor || this.bgcolor;
        this.cornerRadius = data.cornerRadius || this.cornerRadius;
        this.translationZ = data.translationZ || this.translationZ;
    };
    NativeShadowDirective.prototype.loadFromIOSData = function (data) {
        this.maskToBounds = data.maskToBounds || this.maskToBounds;
        this.shadowColor = data.shadowColor || this.shadowColor;
        this.shadowOffset = data.shadowOffset || this.shadowOffset;
        this.shadowOpacity = data.shadowOpacity || this.shadowOpacity;
        this.shadowRadius = data.shadowRadius || this.shadowRadius;
    };
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "shadow", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "elevation", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], NativeShadowDirective.prototype, "shape", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], NativeShadowDirective.prototype, "bgcolor", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "cornerRadius", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "translationZ", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Boolean)
    ], NativeShadowDirective.prototype, "maskToBounds", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", String)
    ], NativeShadowDirective.prototype, "shadowColor", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "shadowOffset", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "shadowOpacity", void 0);
    __decorate([
        core_1.Input(),
        __metadata("design:type", Object)
    ], NativeShadowDirective.prototype, "shadowRadius", void 0);
    __decorate([
        core_1.HostListener('loaded'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NativeShadowDirective.prototype, "onLoaded", null);
    NativeShadowDirective = __decorate([
        core_1.Directive({ selector: '[shadow]' }),
        __metadata("design:paramtypes", [core_1.ElementRef])
    ], NativeShadowDirective);
    return NativeShadowDirective;
}());
exports.NativeShadowDirective = NativeShadowDirective;
